scripts/esrupgrades/esr.5.37-7.8.mysql.up.sql

```diff
@@ -1,1295 +1,1376 @@
    This part of the migration #41 adds such index */
+/* ==> mysql/000075_alter_upload_sessions_index.up.sql <== */
+/* ==> mysql/000090_create_enums.up.sql <== */
+DELIMITER //
+CREATE PROCEDURE MigrateUploadSessions ()
+BEGIN
+	-- 'CREATE INDEX idx_uploadsessions_type ON UploadSessions(Type);'
+	DECLARE CreateIndex BOOLEAN;
+	DECLARE CreateIndexQuery TEXT DEFAULT NULL;
 
-SET @preparedStatement = (SELECT IF(
-    (
-        SELECT COUNT(*) FROM INFORMATION_SCHEMA.STATISTICS
-        WHERE table_name = 'UploadSessions'
-        AND table_schema = DATABASE()
-        AND index_name = 'idx_uploadsessions_type'
-    ) > 0,
-    'SELECT 1',
-    'CREATE INDEX idx_uploadsessions_type ON UploadSessions(Type);'
-));
-
-PREPARE createIndexIfNotExists FROM @preparedStatement;
-EXECUTE createIndexIfNotExists;
-DEALLOCATE PREPARE createIndexIfNotExists;
+	-- 'DROP INDEX idx_uploadsessions_user_id ON UploadSessions; CREATE INDEX idx_uploadsessions_user_id ON UploadSessions(UserId);'
+	DECLARE AlterIndex BOOLEAN;
+	DECLARE AlterIndexQuery TEXT DEFAULT NULL;
 
-/* ==> mysql/000054_create_crt_channelmembership_count.up.sql <== */
-/* fixCRTChannelMembershipCounts fixes the channel counts, i.e. the total message count,
-total root message count, mention count, and mention count in root messages for users
-who have viewed the channel after the last post in the channel */
+	-- 'ALTER TABLE UploadSessions MODIFY COLUMN Type ENUM("attachment", "import");'
+	DECLARE AlterColumn BOOLEAN;
+	DECLARE AlterColumnQuery TEXT DEFAULT NULL;
 
-DELIMITER //
-CREATE PROCEDURE MigrateCRTChannelMembershipCounts ()
-BEGIN
-	IF(
-		SELECT
-			EXISTS (
-			SELECT
-				* FROM Systems
-			WHERE
-				Name = 'CRTChannelMembershipCountsMigrationComplete') = 0) THEN
-		UPDATE
-			ChannelMembers
-			INNER JOIN Channels ON Channels.Id = ChannelMembers.ChannelId SET
-				MentionCount = 0, MentionCountRoot = 0, MsgCount = Channels.TotalMsgCount, MsgCountRoot = Channels.TotalMsgCountRoot, LastUpdateAt = (
-				SELECT
-					(SELECT ROUND(UNIX_TIMESTAMP(NOW(3))*1000)))
-	WHERE
-		ChannelMembers.LastViewedAt >= Channels.LastPostAt;
-		INSERT INTO Systems
-			VALUES('CRTChannelMembershipCountsMigrationComplete', 'true');
+	SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'UploadSessions'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_uploadsessions_type'
+		INTO CreateIndex;
+
+	SELECT IFNULL(GROUP_CONCAT(column_name ORDER BY seq_in_index), '') = 'Type' FROM information_schema.statistics
+		WHERE table_name = 'UploadSessions'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_uploadsessions_user_id'
+		GROUP BY index_name
+		INTO AlterIndex;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE table_name = 'UploadSessions'
+		AND table_schema = DATABASE()
+		AND column_name = 'Type'
+		AND REPLACE(LOWER(column_type), '"', "'") != "enum('attachment','import')"
+		INTO AlterColumn;
+
+	IF CreateIndex THEN
```

Comment:
> Is there an opportunity to test these conditionals -- I worry there's complexity here that we won't know is broken until we get a customer that runs into it.

The happy path (the conditionals are true when they need to be) is already tested with the diff check. But it's true that we don't test that they are not added when it's not strictly needed. https://mattermost.atlassian.net/browse/MM-51840 will help with testing that the script is useful for newer versions (in the migration path), but we would still need to test that the query built is not only correct but minimal.

I think I'll give it a second thought when I get to that ticket. Sounds good?

Reply:
> Alternatively -- and I apologize if this idea comes too late... -- should we consider a version of this script that assumes a schema that /exactly/ matches the starting schema, making the conditionals altogether unnecessary? We could even make that a condition of using these scripts, e.g. dump your schema and compare before running.

That's a good idea, indeed. I actually started this whole effort with that idea in mind, since it was way easier, but then I though that in case something went wrong, customers may want to run the script more than once. In any case, it's something we can test as well. I created a ticket to investigate: https://mattermost.atlassian.net/browse/MM-51853